src/compiler/checker.ts

```diff
@@ -4110,7 +4132,22 @@ namespace ts {
                 }
 
                 if (minArgumentCount < 0) {
-                    minArgumentCount = declaration.parameters.length;
+                    minArgumentCount = declaration.parameters.length - (thisType ? 1 : 0);
+                }
+                if (!thisType && compilerOptions.strictThis) {
+                    if (declaration.kind === SyntaxKind.FunctionDeclaration
+                        || declaration.kind === SyntaxKind.CallSignature
+                        || declaration.kind == SyntaxKind.FunctionExpression
+                        || declaration.kind === SyntaxKind.FunctionType) {
+                        thisType = voidType;
+                    }
+                    else if ((declaration.kind === SyntaxKind.MethodDeclaration || declaration.kind === SyntaxKind.MethodSignature)
+                        && (isClassLike(declaration.parent) || declaration.parent.kind === SyntaxKind.InterfaceDeclaration)) {
+                        thisType = declaration.flags & NodeFlags.Static ?
+                            getWidenedType(checkExpression((<ClassLikeDeclaration>declaration.parent).name)) :

```

Comment:
Yikes, this `checkExpression` call looks all wrong. The `name` of a class declaration isn't an expression. You need to use `getTypeOfSymbol(getSymbolOfNode(declaration.parent))`.